cc wrapper script not detecting cpp files/adding c++ standard includes without '-xc++'

## Describe the bug

The cc-wrapper script won't add the c++ standard include paths when compiling with `*.cpp/*.cxx/...` files and without `-xc++`. This breaks for example the `clang` package when used with build systems that expect it to detect c++ source files, i. e: `xmake`.

## Steps To Reproduce

Steps to reproduce the behavior:

1. install the clang package
2. try to compile a simple `test.cpp` with `clang test.cpp`
3. clang complains about missing c++ standard headers

## Expected behavior
It finds the standard headers and fails to link to a valid executable due to not linking with libstdc++/libc++. e.g gcc follows the expected behaviour, it fails to link:
```console
[nix-shell:~/]$ gcc test.cc 
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: /tmp/nix-shell-55971-0/ccGMZAHj.o: in function `main':
test.cc:(.text.startup+0x10): undefined reference to `std::cout'
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: test.cc:(.text.startup+0x15): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::operator<< <std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&, char const*)'
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: test.cc:(.text.startup+0x3e): undefined reference to `std::ostream::put(char)'
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: test.cc:(.text.startup+0x46): undefined reference to `std::ostream::flush()'
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: test.cc:(.text.startup+0x57): undefined reference to `std::ctype<char>::_M_widen_init() const'
/nix/store/pbixrqd120nnw8a0ygfpm3mnwlr1mbj7-binutils-2.43.1/bin/ld: test.cc:(.text.startup+0x81): undefined reference to `std::__throw_bad_cast()'
```

## Additional context

How the issue was found: xmake-io/xmake#5842

## Metadata

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.7, NixOS, 24.11 (Vicuna), 24.11.20241111.dc460ec`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - nixpkgs: `/nix/store/4r8s42c9mwfvgdlxv0izb1cmlzrsb5nz-source`
```

## Notify maintainers

@Ericson2314
@RossComputerGuy 
@lovek323 
@dtzWill 
@alyssais 
@rrbutani 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/build-support/cc-wrapper/cc-wrapper.sh
+++ b/pkgs/build-support/cc-wrapper/cc-wrapper.sh
@@ -56,6 +56,9 @@ while (( "$n" < "$nParams" )); do
                 c++*) isCxx=1 ;;
             esac
             ;;
+        # check common c++ file extensions
+        # taken from clang: https://github.com/llvm/llvm-project/blob/4d6a5fc702e568b0456c4d8f9e2307eb6d81e955/clang/lib/Driver/Types.cpp#L299-L369
+        *.C|*.H|*.cc|*.CC|*.cp|*.hh|*.ccm|*.cpp|*.CPP|*.c++|*.C++|*.cxx|*.CXX|*.hpp|*.hxx|*.c++m|*.cppm|*.cxxm) isCxx=1 ;;
         --) # Everything else is positional args!
             # See: https://github.com/llvm/llvm-project/commit/ed1d07282cc9d8e4c25d585e03e5c8a1b6f63a74
 
@@ -65,6 +68,12 @@ while (( "$n" < "$nParams" )); do
 
             positionalArgs=("${params[@]:$n}")
             params=("${params[@]:0:$((n - 1))}")
+
+            for pa in "${positionalArgs[@]}"; do
+                case $pa in
+                    *.C|*.H|*.cc|*.CC|*.cp|*.hh|*.ccm|*.cpp|*.CPP|*.c++|*.C++|*.cxx|*.CXX|*.hpp|*.hxx|*.c++m|*.cppm|*.cxxm) isCxx=1 ;;
+                esac
+            done
             break;
             ;;
         -?*) ;;

